### PR TITLE
Add a new Guess-my-number example.

### PR DESCRIPTION
**Pull request by casavaca**

### CODE REVIEW

**Review by lovettchris — CHANGES_REQUESTED** [review]
Awesome thanks, I would love some word smithing which I can do or if you want to do them let me know and I'll complete the review here.  I'm also thinking it would be great if there was a while loop until the user guesses the right answer.  Also since you are using a do block, can the match Ord.compare become an if/then/else?

File: GuessMyNumber/Main.lean [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import Std.Data
+
+-- Generate a "random" number in [0, 99]
+-- It's not actually random. We'll talk about it later.
+def GetSecret : IO Nat :=
+  IO.rand 0 99
+
+partial def guess (secret : Nat) (prompt : String) : IO Unit := do
+  IO.println prompt
+  let stdin := (←IO.getStdin)
```
Comment (leodemoura) [inline GuessMyNumber/Main.lean:10]: This is correct, but most Lean programmers would write
```lean
  let stdin ← IO.getStdin
```
The same pattern occurs in other places.

File: GuessMyNumber/Main.lean [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  let str   := (←stdin.getLine)
+
+  if str.length == 0 then -- handles eof.
+    return
+
+  let num   := str.trim.toNat?
+  match num with
```
Comment (leodemoura) [inline GuessMyNumber/Main.lean:17]: You may consider writing 
```lean
match str.trim.toNat? with
...
```

File: GuessMyNumber/Main.lean [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  | none   => guess secret "Please enter a valid number"
+  | some i =>
+    match (Ord.compare i secret) with
+    | Ordering.eq => IO.println "It's correct!"
```
Comment (leodemoura) [inline GuessMyNumber/Main.lean:21]: This is a good example to demo one of our new features. You can write this match expression as 
```lean
    match compare i secret with
    | .eq => IO.println "It's correct!"
    | .lt => guess secret "Too small"
    | .gt => guess secret "Too large"
```

**Comment by leodemoura** [comment]
Thanks for submitting the PR. I have only a few minor comments.

File: GuessMyNumber/Main.lean [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import Std.Data
+
+-- Generate a "random" number in [0, 99]
+-- It's not actually random. We'll talk about it later.
+def GetSecret : IO Nat :=
```
Comment (Kha) [inline GuessMyNumber/Main.lean:5]:
```suggestion
def getSecret : IO Nat :=
```

File: GuessMyNumber/Main.lean [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  let num   := str.trim.toNat?
+  match num with
+  | none   => guess secret "Please enter a valid number"
+  | some i =>
+    match (Ord.compare i secret) with
+    | Ordering.eq => IO.println "It's correct!"
+    | Ordering.lt => guess secret "Too small"
+    | Ordering.gt => guess secret "Too large"
+
+partial def main : IO Unit := do
```
Comment (Kha) [inline GuessMyNumber/Main.lean:25]:
```suggestion
def main : IO Unit := do
```

File: GuessMyNumber/Readme.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+```lean
+import Std.Data
+
+-- Generate a "random" number in [0, 99]
+-- It's not actually random. We'll talk about it later.
+def GetSecret : IO Nat :=
+  IO.rand 0 99
+```
+
+First, `IO.rand` is not actually a random number. It's a pseudo random number generator **with the same seed every time**! But that's not important for this discussios. Let's assume it returns a number randomly.
```
Comment (Kha) [inline GuessMyNumber/Readme.md:16]: @leodemoura We should probably move this to the `IO.stdGenRef` initializer https://github.com/leanprover/lean4/blob/8649483b416ec9d919c380d657ceedf8184ed527/src/Lean/Server/FileWorker.lean#L494-L495

File: GuessMyNumber/Main.lean (hunk 1 above)
Comment (lovettchris) [reply GuessMyNumber/Main.lean:10]: let's show both forms here for additional educational value.

**Comment by lovettchris** [comment]
@leodemoura, I will merge this PR so that I can work on the changes proposed by you and Sebastian and do some word smithing.  I think Casavanca got busy with other stuff...

**Comment by leodemoura** [comment]
> @leodemoura, I will merge this PR so that I can work on the changes proposed by you and Sebastian and do some word smithing. I think Casavanca got busy with other stuff...

Sure.

BTW, I just pushed the change @Kha suggested above for the Lean 4 repo.

**Comment by Kha** [comment]
@lovettchris FYI, since you have push access you should be able to push changes directly to the PR branch without merging it. Not that it matters much in this case of course since this repo is not exactly high-traffic.

**Comment by lovettchris** [comment]
> @lovettchris FYI, since you have push access you should be able to push changes directly to the PR branch without merging it. Not that it matters much in this case of course since this repo is not exactly high-traffic.

really, does that work with forks? the code lives here: https://github.com/casavaca/lean4-samples

File: GuessMyNumber/Main.lean (hunk 2 above)
Comment (lovettchris) [reply GuessMyNumber/Main.lean:17]: fixed

File: GuessMyNumber/Main.lean (hunk 3 above)
Comment (lovettchris) [reply GuessMyNumber/Main.lean:21]: Done

File: GuessMyNumber/Main.lean (hunk 4 above)
Comment (lovettchris) [reply GuessMyNumber/Main.lean:5]: fixed

File: GuessMyNumber/Main.lean (hunk 5 above)
Comment (lovettchris) [reply GuessMyNumber/Main.lean:25]: Why? Is it not required on main for some reason?

Comment (leodemoura) [reply GuessMyNumber/Main.lean:25]: `main` is not recursive. We only need to use `partial` if the function is recursive, Lean did not manage to prove its termination automatically, and we don't want to prove it ourselves.

File: GuessMyNumber/Main.lean (hunk 1 above)
Comment (leodemoura) [reply GuessMyNumber/Main.lean:10]: There is no reason for using `let stdin := (←IO.getStdin)`, it is a bad style.

Comment (lovettchris) [reply GuessMyNumber/Main.lean:10]: Oh, interesting.  If it is "bad" why doesn't Lean remove it?  What bad things can happen?  Or are you just saying it is not the preferred way in a "lint" sort of "coding guidlines" kind of way?

File: GuessMyNumber/Main.lean (hunk 5 above)
Comment (lovettchris) [reply GuessMyNumber/Main.lean:25]: Got it - so lack of termination doesn't bubble up to the caller then?

File: GuessMyNumber/Readme.md (hunk 6 above)
Comment (lovettchris) [reply GuessMyNumber/Readme.md:16]: Fixed, thanks.

**Comment by lovettchris** [comment]
This is now pushed and fully edited, see https://github.com/leanprover/lean4-samples/tree/main/GuessMyNumber 

And I enabled CI test runs which all pass:
https://github.com/leanprover/lean4-samples/actions/runs/2573009151

File: GuessMyNumber/Main.lean (hunk 1 above)
Comment (leodemoura) [reply GuessMyNumber/Main.lean:10]: > Oh, interesting. If it is "bad" why doesn't Lean remove it?

It is not incorrect to write `let stdin := (←IO.getStdin)`. The program will behave as expected.

> Or are you just saying it is not the preferred way in a "lint" sort of "coding guidlines" kind of way?

Yes, this is what I mean by "bad style". In the future, someone may write a linter for reporting a warning when someone uses this style, but this is super low-priority. We should just make sure documentation and our examples do not use "bad style".

Comment (lovettchris) [reply GuessMyNumber/Main.lean:10]: Well in my updated version of this (see ) I'm making this an educational moment by writing:

```lean
let stdin := (←IO.getStdin)
let str ← stdin.getLine
```

The `←` notation is just to get things out of the IO box. The type of `stdin.getLine` is `IO
String`, and `str` is of type `String`. Notice the second let here uses a short form of assignment
without the need for the `:=` operator and this is the preferred way of doing it.


Is this ok?

**Comment by leodemoura** [comment]
@lovettchris The following workaround with seeds is not needed anymore. I fixed this issue in the commit I linked above.
```lean
-- Generate a "random" number in [0, 99]
-- It's not actually random. We'll talk about it later.
def getSecret : IO Nat :=
  let seed ← (UInt64.toNat ∘ ByteArray.toUInt64LE!) <$> IO.getRandomBytes 8
  IO.setRandSeed seed
  IO.rand 0 99
```
The simpler version should be enough now: 
```lean
-- Generate a "random" number in [0, 99]
def getSecret : IO Nat :=
  IO.rand 0 99
```

File: GuessMyNumber/Main.lean (hunk 1 above)
Comment (leodemoura) [reply GuessMyNumber/Main.lean:10]: I don't like it.  Note that `let str ← stdin.getLine` is **not** a syntax sugar for `let str := (← stdin.getLine)`. The first one is the "macro". 
```lean
let stdin := (←IO.getStdin)
```
is expanded to
```lean
let aux ← IO.getStdin
let stdin := aux
```
I think it is better to just write code as everybody does it, and let David's book explain monadic code and nested actions.

File: GuessMyNumber/Main.lean (hunk 5 above)
Comment (leodemoura) [reply GuessMyNumber/Main.lean:25]: No, it does not. `partial` affects how Lean compiles the given declaration before sending it to the kernel.

**Pull request merged**